src/passes/OptimizeInstructions.cpp

```diff
@@ -2365,6 +2377,13 @@ struct OptimizeInstructions
       curr->type = Type::i32;
       return Builder(*getModule()).makeUnary(ExtendUInt32, curr);
     }
+    // i64(x) & 0x00000000FFFFFFFF   ==>   i64(i32(x))
+    if (matches(curr,
+                binary(And, any(&left), i64(int64_t(0x00000000FFFFFFFF))))) {
+      Builder builder(*getModule());
+      return builder.makeUnary(ExtendUInt32,
+                               builder.makeUnary(WrapInt64, left));
+    }
```

Comment:
It seems better move this to `FinalOptimizer` due to it could cover this case `(x & 0x00000000FFFFFFFF) & C` and optimize it to `extend(wrap(x)) & C` instead `x & C`